src/main/scala/firrtl/passes/wiring/WiringUtils.scala

```diff
@@ -84,86 +97,132 @@ object WiringUtils {
     childrenMap
   }
 
-  /** Counts the number of instances of a module declared under a top module
-    */
-  def countInstances(childrenMap: ChildrenMap, top: String, module: String): Int = {
-    if(top == module) 1
-    else childrenMap(top).foldLeft(0) { case (count, (i, child)) =>
-      count + countInstances(childrenMap, child, module)
-    }
-  }
-
   /** Returns a module's lineage, containing all children lineages as well
     */
   def getLineage(childrenMap: ChildrenMap, module: String): Lineage =
     Lineage(module, childrenMap(module) map { case (i, m) => (i, getLineage(childrenMap, m)) } )
 
-  /** Sets the sink, sinkParent, source, and sourceParent fields of every
-    *  Lineage in tree
+  /** Return a map of sink instances to source instances that minimizes
+    * distance
+    *
+    * @param sinks a sequence of sink modules
+    * @param source the source module
+    * @param i a graph representing a circuit
+    * @return a map of sink instance names to source instance names
+    * @throws WiringException if a sink is equidistant to two sources
     */
-  def setFields(sinks: Set[String], source: String)(lin: Lineage): Lineage = lin map setFields(sinks, source) match {
-    case l if sinks.contains(l.name) => l.copy(sink = true)
-    case l => 
-      val src = l.name == source
-      val sinkParent = l.children.foldLeft(false) { case (b, (i, m)) => b || m.sink || m.sinkParent }
-      val sourceParent = if(src) true else l.children.foldLeft(false) { case (b, (i, m)) => b || m.source || m.sourceParent }
-      l.copy(sinkParent=sinkParent, sourceParent=sourceParent, source=src)
-  }
+  def sinksToSources(sinks: Seq[Named], source: String, i: InstanceGraph):
+      Map[Seq[WDefInstance], Seq[WDefInstance]] = {
+    val indent = "  "
+
+    val owners = new mutable.HashMap[Seq[WDefInstance], Seq[Seq[WDefInstance]]]
+      .withDefaultValue(Seq())
+    val queue = new mutable.Queue[Seq[WDefInstance]]
+    val visited = new mutable.HashMap[Seq[WDefInstance], Boolean]
+      .withDefaultValue(false)
+
+    i.fullHierarchy.keys
+      .filter { case WDefInstance(_, _, module, _) => module == source }
+      .map    { k => i.fullHierarchy(k)
+        .map  { l => queue.enqueue(l)
+                     owners(l) = Seq(l)                                }}
+
+    while (queue.nonEmpty) {
+      val u = queue.dequeue
```

Comment:
I think this is better as just
```scala
for (u <- queue) { ...
```

Reply:
Updated in a newer version. Not fantastic still, but should be more legible.